# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [6]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/5/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/5/"
    },
    "https://swapi.dev/api/people/3/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/3/"
    },
    "https://swapi.dev/api/planets/1/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/1/"
    },
    "https://swapi.dev/api/starships/9/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/9/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: Davi lopes goncalves
- Integrante 2:
- Integrante 3 (se houver):
- Integrante 4 (se houver):

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 5

**Tarefa:** execute o script e observe:
- status code :  150
- nome :Leia Organa
- altura : 49
- cor dos olhos : brown

In [7]:
url = "https://swapi.dev/api/people/5/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Leia Organa",
    "height": "150",
    "mass": "49",
    "hair_color": "brown",
    "skin_color": "light",
    "eye_color": "brown",
    "birth_year": "19BBY",
    "gender": "female",
    "homeworld": "https://swapi.dev/api/planets/2/",
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/"
    ],
    "species": [],
    "vehicles": [
        "https://swapi.dev/api/vehicles/30/"
    ],
    "starships": [],
    "created": "2014-12-10T15:20:09.791000Z",
    "edited": "2014-12-20T21:17:50.315000Z",
    "url": "https://swapi.dev/api/people/5/"
}


**Respostas do aluno — Personagem 3**

- Status code: 200
- Nome do personagem:R2-D2
- Altura:96
- Cor dos olhos:vermelho
- O JSON parece completo? Sim, o JSON está completo, incluindo informações sobre espécie droide e filmes em que aparece


### 1.2 Personagem 3

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [9]:
url = "https://swapi.dev/api/people/1/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Luke Skywalker",
    "height": "172",
    "mass": "77",
    "hair_color": "blond",
    "skin_color": "fair",
    "eye_color": "blue",
    "birth_year": "19BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/6/"
    ],
    "species": [],
    "vehicles": [
        "https://swapi.dev/api/vehicles/14/",
        "https://swapi.dev/api/vehicles/30/"
    ],
    "starships": [
        "https://swapi.dev/api/starships/12/",
        "https://swapi.dev/api/starships/22/"
    ],
    "created": "2014-12-09T13:50:51.644000Z",
    "edited": "2014-12-20T21:17:56.891000Z",
    "url": "https://swapi.dev/api/people/1/"
}


**Respostas do aluno — Personagem 1**

- Status code: 200
- Nome do personagem:Luke Skywalker
- Altura:172
- Cor dos olhos:blond
- O JSON parece completo?Sim, o JSON traz todas as informações esperadas, incluindo período de rotação, diâmetro, residentes e filmes

### 1.3 Planeta 1

**Tarefa:** execute o bloco e observe:
- nome do planeta :Tatooine
- clima : árido
- terreno :deserto
- população :200000

In [11]:
url = "https://swapi.dev/api/planets/1/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Tatooine",
    "rotation_period": "23",
    "orbital_period": "304",
    "diameter": "10465",
    "climate": "arid",
    "gravity": "1 standard",
    "terrain": "desert",
    "surface_water": "1",
    "population": "200000",
    "residents": [
        "https://swapi.dev/api/people/1/",
        "https://swapi.dev/api/people/2/",
        "https://swapi.dev/api/people/4/",
        "https://swapi.dev/api/people/6/",
        "https://swapi.dev/api/people/7/",
        "https://swapi.dev/api/people/8/",
        "https://swapi.dev/api/people/9/",
        "https://swapi.dev/api/people/11/",
        "https://swapi.dev/api/people/43/",
        "https://swapi.dev/api/people/62/"
    ],
    "films": [
        "https://swapi.dev/api/films/1/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-09T13

**Respostas do aluno — Planeta 1**

- Status code:200
- Nome do planeta:Tatooine
- Clima:arido
- Terreno:deserto
- População:200000
- O JSON parece completo?Sim, o JSON traz todas as informações esperadas, incluindo período de rotação, diâmetro, residentes e filmes

### 1.4 Nave 10

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [13]:
url = "https://swapi.dev/api/starships/9"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Death Star",
    "model": "DS-1 Orbital Battle Station",
    "manufacturer": "Imperial Department of Military Research, Sienar Fleet Systems",
    "cost_in_credits": "1000000000000",
    "length": "120000",
    "max_atmosphering_speed": "n/a",
    "crew": "342,953",
    "passengers": "843,342",
    "cargo_capacity": "1000000000000",
    "consumables": "3 years",
    "hyperdrive_rating": "4.0",
    "MGLT": "10",
    "starship_class": "Deep Space Mobile Battlestation",
    "pilots": [],
    "films": [
        "https://swapi.dev/api/films/1/"
    ],
    "created": "2014-12-10T16:36:50.509000Z",
    "edited": "2014-12-20T21:26:24.783000Z",
    "url": "https://swapi.dev/api/starships/9/"
}


**Respostas do aluno — Nave 9**

- Status code:200
- Nome da nave:Death Star
- Modelo:DS-1 Orbital Battle Station
- Fabricante:Imperial Department of Military Research, Sienar Fleet Systems
- Velocidade máxima em atmosfera:1000000000000
- O JSON parece completo?Sim, o JSON contém todos os detalhes técnicos esperados

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [14]:
url = "https://swapi.dev/api/people/4/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Darth Vader
Altura: 202
Cor dos olhos: yellow
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome? dados["name"]
2. Qual campo foi usado para mostrar a altura?dados["height"]
3. Se você fosse QA, quais desses campos validaria primeiro?
O campo nome  é o identificador principal do personagem
O campo altura  pois é uma informação numérica que pode ser verificada Darth Vader é alto nos filmes
O campo cor dos olhos - para confirmar se está correto

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [15]:
url = "https://swapi.dev/api/planets/9/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Coruscant
Clima: temperate
Terreno: cityscape, mountains
População: 1000000000000


**Perguntas**
1. O campo população parece numérico?Sim, o valor "1000000000000" (1 trilhão) é numérico, mas vem como texto no JSON.
2. O clima foi retornado corretamente?Sim, "temperado" está correto para Coruscant.
3. Se você fosse QA, o que validaria nesse planeta?
Se o nome Coruscant está correto capital da República/Império
Se a população de 1 trilhão faz sentido para uma cidade planeta
Se o terreno paisagem urbana reflete a cidade planeta

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [16]:
url = "https://swapi.dev/api/people/20/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica?Verifica se a requisição HTTP foi bem-sucedida código 200 OK, indicando que o servidor respondeu corretamente
2. O que a validação 2 verifica?Verifica se o campo nome existe no JSON retornado, garantindo que a estrutura de dados contém este campo obrigatório
3. O que a validação 3 verifica?Verifica se o nome não é um texto vazio, assegurando que o campo contém um valor válido Yoda não pode ter nome vazio
4. O que a validação 4 verifica?Verifica se o campo altura existe no JSON, confirmando que a informação de altura está presente importante para personagem pequeno

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [17]:
url = "https://swapi.dev/api/starships/12/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [18]:
url = "https://swapi.dev/api/people/2/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [23]:
url = "https://swapi.dev/api/starships/32/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")

Todos os testes da nave passaram!


## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [22]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado?404
2. Esse comportamento está correto?{"detail": "Not found"}
3. Por que um QA deve testar cenários negativos?Para garantir que a API trata erros adequadamente, retorna mensagens claras e não quebra o sistema quando recebe requisições inválidas. Isso também testa a robustez da aplicação cliente e a experiência do usuário em casos de erro

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [24]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/planets/8/")

Origem dos dados: online
Status Code: 200
name: Naboo
rotation_period: 26
orbital_period: 312
diameter: 12120
climate: temperate
gravity: 1 standard
terrain: grassy hills, swamps, forests, mountains
surface_water: 12
population: 4500000000
residents: ['https://swapi.dev/api/people/3/', 'https://swapi.dev/api/people/21/', 'https://swapi.dev/api/people/35/', 'https://swapi.dev/api/people/36/', 'https://swapi.dev/api/people/37/', 'https://swapi.dev/api/people/38/', 'https://swapi.dev/api/people/39/', 'https://swapi.dev/api/people/42/', 'https://swapi.dev/api/people/60/', 'https://swapi.dev/api/people/61/', 'https://swapi.dev/api/people/66/']
films: ['https://swapi.dev/api/films/3/', 'https://swapi.dev/api/films/4/', 'https://swapi.dev/api/films/5/', 'https://swapi.dev/api/films/6/']
created: 2014-12-10T11:52:31.066000Z
edited: 2014-12-20T20:58:18.430000Z
url: https://swapi.dev/api/planets/8/


**Respostas finais**
1. Qual endpoint você testou no desafio?https://swapi.dev/api/planets/8/
2. Qual foi o status code?200
3. Qual campo chamou mais a sua atenção?clima temperado, como visto nos filmes
4. Se você fosse QA, quais 3 validações faria nesse recurso?Validar se o clima "temperado" corresponde ao mostrado nos filmes
Verificar se os residentes listados realmente são de Naboo Padmé, Palpatine
Confirmar se o terreno "selvas, oceanos" está correto Naboo tem água e florestas

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
